agent/fw_drupal8.c

```diff
@@ -605,12 +605,128 @@ NR_PHP_WRAPPER(nr_drupal94_invoke_all_with_clean) {
 NR_PHP_WRAPPER_END
 #endif  // OAPI
 
+#define NR_FREE_HOOK_MEM \
+  nr_free(hook_str);     \
+  nr_free(class_str);    \
+  nr_free(method_str);   \
+  nr_free(module_str);   \
+  nr_free(hookpath);
+
+/*
+ * Purpose: Instrument Drupal Attribute Hooks for Drupal 11.1+
+ *
+ * Params: 1. A zval pointer to the moduleHandler instance in use by Drupal.
+ *
+ * Return: bool
+ *
+ */
+static bool nr_drupal_hook_attribute_instrument(zval* module_handler) {
+  zval* hook_implementation_map = NULL;
+
+  nr_php_string_hash_key_t* hook_key = NULL;
+  zval* hook_val = NULL;
+  nr_php_string_hash_key_t* class_key = NULL;
+  zval* class_val = NULL;
+  nr_php_string_hash_key_t* method_key = NULL;
+  zval* module_val = NULL;
+  zend_ulong key_num = 0;
+
+  char* hook_str = NULL;
+  char* class_str = NULL;
+  char* method_str = NULL;
+  char* module_str = NULL;
+  char* hookpath = NULL;
+
+  hook_implementation_map = nr_php_get_zval_object_property(
+      module_handler, "hookImplementationsMap");
+  if (hook_implementation_map) {
+    if (nr_php_is_zval_valid_array(hook_implementation_map)) {
+      ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(hook_implementation_map), key_num,
+                                hook_key, hook_val) {
+        (void)key_num;
+        if ((NULL == hook_key) || (0 == nr_php_is_zval_valid_array(hook_val))) {
+          nrl_warning(NRL_FRAMEWORK,
+                      "hookImplementationsMap[hook = %s]: invalid value",
+                      NRSAFESTR(ZEND_STRING_VALUE(hook_key)));
+          NR_FREE_HOOK_MEM
+          return false;
+        }
+
+        ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(hook_val), key_num, class_key,
+                                  class_val) {
+          (void)key_num;
+          if ((NULL == class_key)
+              || (0 == nr_php_is_zval_valid_array(class_val))) {
+            nrl_warning(NRL_FRAMEWORK,
+                        "hookImplementationsMap[class = %s]: invalid value",
+                        NRSAFESTR(ZEND_STRING_VALUE(class_key)));
+            NR_FREE_HOOK_MEM
+            return false;
+          }
+
+          ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(class_val), key_num, method_key,
+                                    module_val) {
+            (void)key_num;
+
+            NR_FREE_HOOK_MEM
```

Comment:
This is confusing. Why is it here and not after when it was used, i.e. after call to `nr_php_wrap_user_function_drupal`?

Reply:
In order to guarantee the memory is free before trying to reassign. In a triple nested loop it's much safer to force free variables at the start than the end in my opinion.

Reply:
Isn't it safer to free them immediately after t[hey are allocated and used and no longer needed after the call to `nr_php_wrap_user_function_drupal`](https://github.com/newrelic/newrelic-php-agent/pull/1030/files#diff-6c636f6e3733aae919f4c4cf66ec5226065346587de8cbaf3918f483f9d0f523R682-R697)?  Why wait for it to loop through again and sprinkle calls to NR_FREE_HOOK_MEM throughout when you can just free [right after using](https://github.com/newrelic/newrelic-php-agent/pull/1030/files#r1972498681)?

Reply:
Changed here 3ded06d732c7cd7b4ee87b62f4d57a3dbbc60ef8

Reply:
3c68c3d7063bf0c0915a551b9c68d289cbb78732